src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SslOverTdsStream.NetCoreApp.cs

```diff
@@ -0,0 +1,289 @@
+        {
+            if (_encapsulate)
+            {
+                if (_packetBytes > 0)
+                {
+                    // there are queued bytes from a previous packet available
+                    // work out how many of the remaining bytes we can consume
+                    int wantedCount = Math.Min(buffer.Length, _packetBytes);
+                    int readCount = _stream.Read(buffer.Slice(0, wantedCount));
+                    if (readCount == 0)
+                    {
+                        // 0 means the connection was closed, tell the caller
+                        return 0;
+                    }
+                    _packetBytes -= readCount;
+                    return readCount;
+                }
+                else
+                {
+                    Span<byte> headerBytes = stackalloc byte[TdsEnums.HEADER_LEN];
+
+                    // fetch the packet header to determine how long the packet is
+                    int headerBytesRead = 0;
+                    do
+                    {
+                        int headerBytesReadIteration = _stream.Read(headerBytes.Slice(headerBytesRead, (TdsEnums.HEADER_LEN - headerBytesRead)));
+                        if (headerBytesReadIteration == 0)
+                        {
+                            // 0 means the connection was closed, tell the caller
+                            return 0;
+                        }
+                        headerBytesRead += headerBytesReadIteration;
+                    } while (headerBytesRead < TdsEnums.HEADER_LEN);
+
+                    // read the packet data size from the header and store it in case it is needed for a subsequent call
+                    _packetBytes = ((headerBytes[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8) | headerBytes[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - TdsEnums.HEADER_LEN;
+
+                    // read as much from the packet as the caller can accept
+                    int packetBytesRead = _stream.Read(buffer.Slice(0, Math.Min(buffer.Length, _packetBytes)));
+                    _packetBytes -= packetBytesRead;
+                    return packetBytesRead;
+                }
+            }
+            else
+            {
+                return _stream.Read(buffer);
+            }
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (_encapsulate)
+            {
+                if (_packetBytes > 0)
+                {
+                    // there are queued bytes from a previous packet available
+                    // work out how many of the remaining bytes we can consume
+                    int wantedCount = Math.Min(buffer.Length, _packetBytes);
+
+                    int readCount;
+                    {
+                        ValueTask<int> remainderReadValueTask = _stream.ReadAsync(buffer.Slice(0, wantedCount), cancellationToken);
```

Comment:
Am curious... What is gained by the split logic for IsCompletedSuccessfully? Why not just always await?

Reply:
It saves a call into the state machine which can avoid a lot of work. it's a common pattern with hot paths and since this is internal and will be required on all network io I thought I may as well include it. You can see it in https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/

Reply:
I'm familiar with the pattern, but only when the async path has some additional logic that you want to skip in the sync path. For example, in the blog post you point to, the async path contains a `RegisterCancellation` invocation which should indeed be skipped if the operation completes synchronously - but here there's no such thing. Note also how Stephen explicitly recommends developers use this pattern "hopefully only after measuring carefully and finding it provides meaningful benefit".

Out of curiosity, I went ahead and measured the difference and got this:

``` ini

BenchmarkDotNet=v0.12.0, OS=ubuntu 19.10
Intel Xeon W-2133 CPU 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.100-preview.2.20176.6
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


```
|      Method |     Mean |    Error |   StdDev |   Median |
|------------ |---------:|---------:|---------:|---------:|
|  SplitLogic | 48.28 ns | 0.976 ns | 1.631 ns | 47.33 ns |
| JustAwaitIt | 55.31 ns | 0.047 ns | 0.042 ns | 55.31 ns |

<details>
 <summary>Benchmark code</summary>

```c#
public class Program
{
    [Benchmark]
    public async ValueTask<int> SplitLogic()
    {
        var valueTask = SomeSyncReturningAsyncMethod();
        return valueTask.IsCompletedSuccessfully
            ? valueTask.Result
            : await SomeSyncReturningAsyncMethod();
    }
    
    [Benchmark]
    public async ValueTask<int> JustAwaitIt()
        => await SomeSyncReturningAsyncMethod();

    ValueTask<int> SomeSyncReturningAsyncMethod() => new ValueTask<int>(8);
    
    static void Main(string[] args)
        => BenchmarkRunner.Run<Program>();
}
```
</details>

I personally wouldn't complicate the code with this pattern for a 7ns improvement - not in this project anyway, where it' almost sure to be lost in overall perf - but of course that's up to you and the SqlClient people to decide on. In any case, when working any kind of micro-optimizations like this one, it's really a good idea to have a BenchmarkDotNet suite to verify that there's meaningful improvement.

Reply:
Thanks for the numbers. It is a tiny change in times. I view this part of the codebase as both performance critical and low level because it blocks any better performance higher in the pipeline so as you've seen I chose to go with the more complex and less maintainable long form. I've no problem changing it if someone strongly desires it.

Reply:
This part of the codebase does indeed seem quite critical for perf and what you're doing seems important (am lacking context here and am just looking at the PR code). My point is only that specifically for this ValueTask pattern, I'd only introduce this kind of code when a functional benchmark shows it has a meaningful impact. My benchmark above isn't that - I'd be extremely surprised if this could be shown to have any impact on actual SqlClient code executing a command (and code complexity does have a price). Up to you guys.